src/NetworkManager.cpp

```diff
@@ -97,24 +106,75 @@ bool NetworkManager::IsUrlAllowed(const std::string& url)
 
 HttpRequestFuturePtr NetworkManager::HttpRequest(const HttpRequestArgs& args)
 {
-	ix::HttpRequestArgsPtr req = this->httpClient.createRequest(args.url, args.method);
+	auto &client = args.downloadFile.empty() ? this->httpClient : this->downloadClient;
+	auto downloadFile = std::make_shared<RageFile>();
+	std::string downloadFilename;
+
+	ix::HttpRequestArgsPtr req = client.createRequest(args.url, args.method);
 	req->body = args.body;
 	req->multipartBoundary = args.multipartBoundary;
 
+	req->extraHeaders["User-Agent"] = this->GetUserAgent();
 	for (const auto& entry : args.headers)
 	{
 		req->extraHeaders[entry.first] = entry.second;
 	}
 
-	if (args.connectTimeout > 0) {
-		req->connectTimeout = args.connectTimeout;
+	if (!args.downloadFile.empty())
+	{
+		downloadFilename = std::string("/Downloads/") + args.downloadFile;
+		if (!downloadFile->Open(downloadFilename, RageFile::WRITE | RageFile::STREAMED))
+		{
+			req->cancel = true;
+		}
+
+		req->onChunkCallback = [args, downloadFile, req](const std::string& data) {
+			if (downloadFile->Write(data.c_str(), data.size()) < 0)
+			{
+				req->cancel = true;
+			}
+		};
+
+		// Don't timeout downloads by default. This value might be
+		// overwritten below if a value is specified in the request.
+		req->transferTimeout = INT_MAX;
+
 	}
 
-	if (args.transferTimeout > 0) {
+	if (args.connectTimeout > 0)
+		req->connectTimeout = args.connectTimeout;
+
+	if (args.transferTimeout > 0)
 		req->transferTimeout = args.transferTimeout;
-	}
 
-	this->httpClient.performRequest(req, args.onResponse);
+	if (args.onProgress)
+		req->onProgressCallback = args.onProgress;
+
+	client.performRequest(req, [args, downloadFile, downloadFilename](const ix::HttpResponsePtr& response) {
+		if (downloadFile->IsOpen())
+		{
+			RString error = downloadFile->GetError();
+			downloadFile->Close();
+
+			if (!error.empty())
+			{
+				std::string errorMessage = "could not write to " + downloadFile->GetPath() + ": " + error;
+				if (args.onFileError)
+					args.onFileError(errorMessage);
+
+				FILEMAN->Remove(downloadFilename);
+				return;
+			}
+		}
+
+		if (args.onResponse)
+			args.onResponse(response);
+
+		if (!args.downloadFile.empty())
+		{
+			FILEMAN->Remove(downloadFilename);
+		}
```

Comment:
The download is removed here. `ClearDownloads()` is only used to clean up downloads during startup after a crash that caused a download to not get cleaned up.

Reply:
Understood!